Fix compilation with -std=c++20 and gcc-10

It is currently not possible to compile deal.II with `-std=c++20` due to new rules regarding operator overloading as evidenced here: https://cdash.43-1.org/viewBuildError.php?buildid=6688

As a temporary workaround it is enough to remove `LinearIndexIterator::operator!=()` which allows the compiler to resolve the overload in some compatibility mode (that simply emits a warning instead).

A proper fix for this issue might be to remove all `operator@@()` variants (in case `DEAL_II_HAVE_CXX20` is true) and replace them by a single `operator<=>()`.

Minimal testcase:
```
template <typename Derived> class Iterator {
public:
  bool operator==(const Derived &other) { return true; }
  bool operator!=(const Derived &other) { return false; }
};

class MyIterator : public Iterator<MyIterator> {};

int main() {
  MyIterator it1, it2;
  it1 == it2;
}
```
This emits a warning:
```
test.cc:11:7: warning: ISO C++20 considers use of overloaded operator '==' (with operand types 'MyIterator' and 'MyIterator') to be ambiguous despite there being a unique best viable function [-Wambiguous-reversed-operator]
  it1 == it2;
  ~~~ ^  ~~~
test.cc:3:8: note: ambiguity is between a regular call to this operator and a call with the argument order reversed
  bool operator==(const Derived &other) { return true; }
       ^
```
Now changing `==` to `!=`  as follows:
```
// ...
int main() {
  MyIterator it1, it2;
  it1 != it2;
}
```
emits an error:
```
test.cc:11:7: error: use of overloaded operator '!=' is ambiguous (with operand types 'MyIterator' and 'MyIterator')
  it1 != it2;
  ~~~ ^  ~~~
test.cc:4:8: note: candidate function
  bool operator!=(const Derived &other) { return false; }
       ^
test.cc:3:8: note: candidate function
  bool operator==(const Derived &other) { return true; }
       ^
test.cc:3:8: note: candidate function (with reversed parameter order)
```

--- a/include/deal.II/base/linear_index_iterator.h
+++ b/include/deal.II/base/linear_index_iterator.h
@@ -249,8 +249,16 @@ public:
    * Comparison operator. Returns <code>true</code> if both iterators point to
    * the same entry in the same container.
    */
-  bool
-  operator==(const LinearIndexIterator &) const;
+  template <typename OtherIterator>
+  friend typename std::enable_if<
+    std::is_convertible<OtherIterator, DerivedIterator>::value,
+    bool>::type
+  operator==(const LinearIndexIterator &left, const OtherIterator &right)
+  {
+    const auto &right_2 = static_cast<const DerivedIterator &>(right);
+    return left.accessor.container == right_2.accessor.container &&
+           left.accessor.linear_index == right_2.accessor.linear_index;
+  }
 
   /**
    * Inverse of operator==().
@@ -440,17 +448,6 @@ inline typename LinearIndexIterator<DerivedIterator, AccessorType>::pointer
   LinearIndexIterator<DerivedIterator, AccessorType>::operator->() const
 {
   return &accessor;
-}
-
-
-
-template <class DerivedIterator, class AccessorType>
-inline bool
-LinearIndexIterator<DerivedIterator, AccessorType>::operator==(
-  const LinearIndexIterator<DerivedIterator, AccessorType> &other) const
-{
-  return accessor.container == other.accessor.container &&
-         accessor.linear_index == other.accessor.linear_index;
 }
 
 

--- a/include/deal.II/base/linear_index_iterator.h
+++ b/include/deal.II/base/linear_index_iterator.h
@@ -250,7 +250,7 @@ public:
    * the same entry in the same container.
    */
   bool
-  operator==(const DerivedIterator &) const;
+  operator==(const LinearIndexIterator &) const;
 
   /**
    * Inverse of operator==().
@@ -446,12 +446,11 @@ inline typename LinearIndexIterator<DerivedIterator, AccessorType>::pointer
 
 template <class DerivedIterator, class AccessorType>
 inline bool
-LinearIndexIterator<DerivedIterator, AccessorType>::
-operator==(const DerivedIterator &other) const
+LinearIndexIterator<DerivedIterator, AccessorType>::operator==(
+  const LinearIndexIterator<DerivedIterator, AccessorType> &other) const
 {
-  const auto &other_2 = static_cast<decltype(*this) &>(other);
-  return accessor.container == other_2.accessor.container &&
-         accessor.linear_index == other_2.accessor.linear_index;
+  return accessor.container == other.accessor.container &&
+         accessor.linear_index == other.accessor.linear_index;
 }
 
 

--- a/include/deal.II/base/linear_index_iterator.h
+++ b/include/deal.II/base/linear_index_iterator.h
@@ -256,8 +256,7 @@ public:
   operator==(const LinearIndexIterator &left, const OtherIterator &right)
   {
     const auto &right_2 = static_cast<const DerivedIterator &>(right);
-    return left.accessor.container == right_2.accessor.container &&
-           left.accessor.linear_index == right_2.accessor.linear_index;
+    return left.accessor == right_2.accessor;
   }
 
   /**

--- a/include/deal.II/base/table.h
+++ b/include/deal.II/base/table.h
@@ -908,6 +908,19 @@ namespace MatrixTableIterators
                  const std::ptrdiff_t         linear_index);
 
     /**
+     * Comparison operator.
+     */
+    template <bool OtherConstness>
+    friend bool
+    operator==(
+      const AccessorBase<TableType, Constness, storage_order> &     left,
+      const AccessorBase<TableType, OtherConstness, storage_order> &right)
+    {
+      return left.container == right.container &&
+             left.linear_index == right.linear_index;
+    }
+
+    /**
      * Get a constant reference to the value of the element represented by
      * this accessor.
      */

--- a/include/deal.II/grid/tria_iterator.h
+++ b/include/deal.II/grid/tria_iterator.h
@@ -379,8 +379,10 @@ public:
   /**
    * Compare for equality.
    */
-  bool
-  operator==(const TriaRawIterator &) const;
+  template <typename OtherAccessor = Accessor>
+  typename std::enable_if<std::is_convertible<OtherAccessor, Accessor>::value,
+                          bool>::type
+  operator==(const TriaRawIterator<OtherAccessor> &) const;
 
   /**
    * Compare for inequality.

--- a/include/deal.II/grid/tria_iterator.templates.h
+++ b/include/deal.II/grid/tria_iterator.templates.h
@@ -82,9 +82,12 @@ TriaRawIterator<Accessor>::operator=(const TriaRawIterator<Accessor> &i)
 
 
 template <typename Accessor>
-inline bool
-TriaRawIterator<Accessor>::
-operator==(const TriaRawIterator<Accessor> &other) const
+template <typename OtherAccessor>
+inline
+  typename std::enable_if<std::is_convertible<OtherAccessor, Accessor>::value,
+                          bool>::type
+  TriaRawIterator<Accessor>::
+  operator==(const TriaRawIterator<OtherAccessor> &other) const
 {
   return accessor == other.accessor;
 }

--- a/include/deal.II/lac/sparsity_pattern.h
+++ b/include/deal.II/lac/sparsity_pattern.h
@@ -1418,8 +1418,6 @@ namespace SparsityPatternIterators
   inline bool
   Accessor::operator==(const Accessor &other) const
   {
-    Assert(container != nullptr, DummyAccessor());
-    Assert(other.container != nullptr, DummyAccessor());
     return (container == other.container && linear_index == other.linear_index);
   }
 

--- a/source/distributed/fully_distributed_tria.cc
+++ b/source/distributed/fully_distributed_tria.cc
@@ -75,13 +75,13 @@ namespace parallel
         this->set_mesh_smoothing(
           static_cast<
             typename dealii::Triangulation<dim, spacedim>::MeshSmoothing>(
-            dealii::Triangulation<dim>::none |
+            dealii::Triangulation<dim, spacedim>::none |
             Triangulation<dim, spacedim>::limit_level_difference_at_vertices));
       else
         this->set_mesh_smoothing(
           static_cast<
             typename dealii::Triangulation<dim, spacedim>::MeshSmoothing>(
-            dealii::Triangulation<dim>::none));
+            dealii::Triangulation<dim, spacedim>::none));
 
       this->set_mesh_smoothing(construction_data.smoothing);
 

